### DISCUSSION TITLE
Prisma multi-tenant features?

### DISCUSSION THREAD

**Discussion by florianmartens** [prisma/prisma#2846] [Q&A] [answered]
Thanks for the work on Prisma.

Does Prisma have any significant multi-tenant features? There are two features I'd really love / need for development:

1.) The ability to annotate all queries for a model e. g. `findOne` `find` with an additional condition. In this case, a query like `find()` would automatically become `findOne({ where: { tenantID } })` - this would allow isolation on the logic layer that is more robust than adding a filter condition to every query.

2.) The ability to pool database connections, allowing me to manage connections to many databases as well as run migrations with them.

Maybe I've overlooked some features.

**Comment by ryands17** [comment, marked as answer]
Hey @florianmartens :wave: 

> 1.) The ability to annotate all queries for a model e. g. findOne find with an additional condition. In this case, a query like find() would automatically become findOne({ where: { tenantID } }) - this would allow isolation on the logic layer that is more robust than adding a filter condition to every query

I would suggest creating the feature request for this [here](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=feature_request.md&title=) with the above use case.

> 2.) The ability to pool database connections, allowing me to manage connections to many databases as well as run migrations with them.

For this, there is an on-going issue [here](https://github.com/prisma/prisma/issues/2077) and someone from the community has also created a library for that which you can check out.

**Comment by pantharshit00** [comment]
@florianmartens Please also checkout this community project which might be interesting to you: https://github.com/errorname/prisma-multi-tenant

**Comment by Errorname** [comment]
Hi @florianmartens 

Currently,  [Prisma-multi-tenant](https://github.com/errorname/prisma-multi-tenant) uses the "one DB per tenant" model of multi-tenancy. It's able to migrate all your databases at once, and more. This should satisfy your second feature request.

Concerning your first feature request, it should now become possible to do this with the [Middleware](https://github.com/prisma/prisma-client-js/issues/770) feature of Prisma. I intend to improve Prisma-multi-tenant to provide a "Single DB for all tenants" mode, which would add the `{ where: { tenantId } }` argument on all your queries. You can follow the issue [here](https://github.com/Errorname/prisma-multi-tenant/issues/40)

Finally, if you want a mix of those two features: "multiple DB, but with multiple tenants on each", there is also [an issue](https://github.com/Errorname/prisma-multi-tenant/issues/31) on Prisma-multi-tenant to implement this model. However, it will take time for me to implement this one I think.

**Reply by SharadKumar** [reply]
Looking forward to this @Errorname !!
Very useful.

**Reply by firmanjml** [reply]
Do we have another approach for this because prisma-multi-tenant seems to be outdated and no longer maintained

**Reply by Jarrodsz** [reply]
Would like to know the same

**Reply by arpanpreneur** [reply]
Yes very useful. For most self-serve kind of B2B applications, tenat_id based shared database makes way more sense. Similar to ActsAsTenant in Rails and django-multitenant (by Citus) in Django.

**Comment by mcelotti** [comment]
Now that prisma offers client extensions (https://www.prisma.io/docs/concepts/components/prisma-client/client-extensions), I'd suggest you to take a look at this https://dev.to/moofoo/nestjspostgresprisma-multi-tenancy-using-nestjs-prisma-nestjs-cls-and-prisma-client-extensions-ok7

**Comment by ludralph** [comment]
Hi @florianmartens,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
